README.md
free-game
=========

[![Build Status](https://secure.travis-ci.org/fumieval/free-game.png?branch=master)](http://travis-ci.org/fumieval/free-game)

free-game is a library that abstracts graphical applications with simple interfaces.

Install
-------------------------------------------------------------------------------------

    $ cabal install free-game


Migration Guide
-------------------------------------------------------------------------------------

* `runGame param`
    * `runGame Window (BoundingBox 0 0 640 480)` or `runGame FullScreen (BoundingBox 0 0 640 480)`
    * `setTitle "Lorem ipsum"`
    * `showCursor` or `hideCursor`
    * `clearColor black`
    * `setFPS 60`

* `loadBitmapsWith 'func "/path/to"`
    * `loadBitmapsWith [|func|] "/path/to"`
* `keySpecial`
    * `keyPress` or `keyDown` or `keyUp` -- You don't have to keep key states anymore!
* `foreverTick m`
    * `foreverFrame m` -- It is faster
* `do { some draw-only computation }`
    * Just apply `draw` to make your code faster.
* `Game` is a kind of coroutine upon `Frame`.

Comparison with [gloss](http://hackage.haskell.org/package/gloss)
-------------------------------------------------------------------------------------
free-game's API is similar to what gloss have.

| gloss         | free-game           |
| ------------- | ------------------- |
| Picture       | Game ()             |
| Blank         | return ()           |
| Text          | text font size      |
| Bitmap w h d  | bitmap              |
| Translate x y | translate (V2 x y)  |
| Scale x y     | scale (V2 x y)      |
| Rotate        | rotate              |
| Color         | color               |
| Pictures xs   | sequence_ xs        |
| Circle        | circle              |
| Line          | line                |
| Polygon       | polygon             |

Special Thanks
------------------------------------------------------------------------------------

* [jbracker](https://github.com/jbracker), extended features and fixed minor bugs

Bug reports, pull requests, feature requests are welcome.

Donate
-------------------------------------

Bitcoin Address: 1EvewG7YHdcgMjmQEYEBrKiPjUwGWBckY4


CHANGELOG
1.0.4
==========================================================================

* Fixed some potential bugs that appeared on 7.8.1 RC2
* Added `getBoundingBox` and `setBoundingBox` which accesses the window size and the region to draw.
* `Resizable`, the new constructor of `WindowMode`, will create a resizable window.
* Demoted the precedence of `thickness` and `blendMode` according to other APIs.

1.0.3
==========================================================================
* Added `runGameDefault` as an alternative of classic `runGame def`.
* Removed the duplicate instance of `MonadIO`.
* `free-game` no longer depends on ominous `repa`.
* Reconstructed 'FreeGame.Data.Bitmap'. 'Bitmap' is just an alias of `Codec.Picture.Repa.Image PixelRGBA8`
* Added `bitmapOnce` which does not keep the internal texture to draw.
* Added `forkFrame` analogous to `forkIO`.
* Accelerate text rendering.
* Make the window size solid.

1.0.2
==========================================================================
* Supported changing a blend function. `blendMode mode m` changes the blend mode while `m` is running.
* Fixed fatal 'keyPress'-related bugs.
* Special thanks: [@myuon_myon](https://twitter.com/myuon_myon)
* Re-added `keyChar` and `keySpecial`.

1.0.1
==========================================================================
* Demoted the precedence of `Affine` APIs to 5.

1.0
==========================================================================
* Supported free-4.4.
* Supported GLFW-b-1.3.
* Use `Double` instead of `Float`.
* Make it more efficient.
* `loadBitmaps` takes an expression instead of a `Name`.
* Reorganized typeclasses.
* Rename: `fromBitmap` -> `bitmap`
* Rename: `colored` -> `color`
* New API: `takeScreenshot`
* New API: `getFPS`, `setFPS`
* Now the verbose module prefix `Graphics.UI` is extinct.
* And a bunch of renovations...

examples/demo.hs
-- Pointless demo
{-# LANGUAGE OverloadedStrings #-}
import FreeGame
import Control.Applicative
import Control.Monad
import Control.Monad.State.Strict

figureTest :: Frame ()
figureTest = draw $ do
    color cyan -- 'colored' gives a color to the action.
        $ line [V2 80 80, V2 160 160] -- 'line' draws line through the given coordinates.

    color green $ polygon [V2 20 0, V2 100 20, V2 90 60, V2 30 70]

    color blue
        $ translate (V2 0 200) -- 'translate' moves the action.
        $ rotateD 45 -- 'rotateD' rotates the action. (in degrees).
        $ scale 2 -- 'scale' resizes the action. You can also specify as (V2 x y).
        $ polygonOutline [V2 20 0, V2 100 20, V2 50 60]

    color magenta
        $ thickness 3 -- 'thickness' sets the thickness of the lines.
        $ translate (V2 100 300)
        $ circleOutline 50

fontTest :: Font -> Frame ()
fontTest font = do
    translate (V2 30 300) $ color white $ do
        text font 48 "The quick brown fox"
        color red $ line [V2 (-100) 0, V2 640 0]
        translate (V2 0 80) $ do
            text font 48 "jumps over the lazy dog"
            color red $ line [V2 (-100) 0, V2 640 0]
        translate (V2 0 160) $ do
            text font 48 "0123456789" -- Use 'text font size string' to draw a string.
            color red $ line [V2 (-100) 0, V2 640 0]
        
        color red $ line [V2 0 (-600), V2 0 600]

bitmapTest :: Bitmap -> Frame ()
bitmapTest bmp = blendMode Add $ do
    
    color (Color 1 0 0 1) $ do
        translate (V2 300 346) $ bitmap bmp -- 'bitmap' creates an action from the bitmap.
    color (Color 0 1 0 1) $ do
        translate (V2 310 350) $ bitmap bmp -- 'bitmap' creates an action from the bitmap.
    color (Color 0 0 1 1) $ do
        translate (V2 293 359) $ bitmap bmp -- 'bitmap' creates an action from the bitmap.
    
mouseTest :: Frame ()
mouseTest = do
    p <- mousePosition
    l <- mouseButtonL
    r <- mouseButtonR
    let col = case (l, r) of
            (False, False) -> black
            (True, False) -> red
            (False, True) -> blue
            (True, True) -> blend 0.5 red blue
    translate p $ color col $ thickness 4 $ circleOutline 16

main = runGame Windowed (BoundingBox 0 0 640 480) $ do
    bmp <- readBitmap "bird.png"
    bmp' <- readBitmap "logo.png"
    font <- loadFont "VL-PGothic-Regular.ttf"
    let bmp' = cropBitmap bmp (128, 128) (64, 64)
    clearColor black
    forkFrame $ preloadBitmap bmp'
    foreverFrame $ do

        bitmapTest bmp
        figureTest
        translate (V2 320 80) $ bitmap bmp' -- You can slice bitmaps using 'cropBitmap'.

        fontTest font
        translate (V2 240 240) $ do
            mouseTest

            fps <- getFPS

            color black $ text font 15 (show fps)

        whenM (keyDown KeyA) $ translate (V2 300 300) $ color black $ text font 30 "A"
        whenM (keyPress KeyA) $ translate (V2 320 300) $ color black $ text font 30 "B"
        whenM (keyUp KeyA) $ translate (V2 340 300) $ color black $ text font 30 "C"
        whenM (keyDown KeyS) $ takeScreenshot >>= writeBitmap "capture.png"
